File: 1 .Linked-List/LinkedList/LinkedList.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace LinkedList
{
    public class LinkedList<T> : IEnumerable<T>
    {
        private class ListNode<T>
        {
            public ListNode(T value)
            {
                this.Value = value;
            }

            public T Value { get; private set; }

            public ListNode<T> NextNode { get; set; }
        }

        private ListNode<T> head;
        private ListNode<T> tail;

        public int Count { get; private set; }

        public void Add(T value)
        {
            var newNode = new ListNode<T>(value);
            if (this.head == null)
            {
                this.head = newNode;
                this.tail = newNode;
            }
            else
            {
                this.tail.NextNode = newNode;
                this.tail = newNode;
            }
            this.Count++;
        }

        public void Remove(int index)
        {
            if (index < 0 || index >= this.Count)
            {
                throw new IndexOutOfRangeException("Ïnvalid Index!");
            }
            if (this.tail == null)
            {
                throw new InvalidOperationException("Empty list");
            }
            if (index == 0)
            {
                this.head = this.head.NextNode;
            }
            else
            {
                var currentNode = this.head;
                for (int i = 1; i < index; i++)
                {
                    currentNode = currentNode.NextNode;
                }
                var prev = currentNode;
                currentNode = currentNode.NextNode;
                var next = currentNode.NextNode;
                if (next == null)
                {
                    this.tail = prev;
                }
                prev.NextNode = next;
            }
            this.Count--;
        }

        public IEnumerator<T> GetEnumerator()
        {
            var currentNode = this.head;
            while (currentNode != null)
            {
                yield return currentNode.Value;
                currentNode = currentNode.NextNode;
            }
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }

    }
}

File: Java/src/linkedList/LinkedList.java
package linkedList;

import java.util.Iterator;

public class LinkedList<T> implements Iterable<T> {

	private Node<T> head;
	private Node<T> tail;
	private int count;

	public int getCount() {
		return this.count;
	}

	public void setCount(int count) {
		this.count = count;
	}

	public void add(T value) {
		Node<T> newNode = new Node<T>(value);

		if (this.head == null) {
			this.head = newNode;
			this.tail = newNode;
		} else {
			this.tail.setNextNode(newNode);
			this.tail = newNode;
		}
		this.count++;
	}

	public T remove(int index) {
		if (index < 0 || index >= this.count) {
			throw new Error("Invalid Index!");
		}
		if (this.count == 0) {
			throw new Error("List is empty!");
		}
		if (index == 0) {
			T result = this.head.getValue();
			this.head = this.head.getNextNode();
			return result;
		} else {
			Node<T> current = this.head;
			for (int i = 1; i < index; i++) {
				current = current.getNextNode();
			}
			Node<T> previous = current;
			current = current.getNextNode();
			T result = current.getValue();
			Node<T> next = current.getNextNode();
			if (next == null) {
				this.tail = previous;
			}
			previous.setNextNode(next);
			return result;
		}
	}

	@Override
	public Iterator<T> iterator() {
		return new LinkedListIterator<T>(this.head);
	}

}
